Add how to store tokens safely post (#3)
BrunoQuaresma/blog: content/blog/how-to-store-tokens-safely/index.md

---
title: How to Store Tokens Safely
date: "2019-10-29"
description: "Let's talk about how to store your tokens safely using cookies and lambdas."
---

> How to store tokens safely? It's a common question on frontend community and there is some alternatives to this. In this post, I'll show you how to do this using cookies and lambdas.

Reviewer: ```suggestion
How should tokens be stored safely? It's a common question within the frontend community, and there are a few alternatives. In this post, I'll show you how to do this using cookies and lambdas.
```

> Before we get started, we need to know use local storage is not a good idea, as malicious scripts can iterate over the local storage and send the info to an external API. So if you have any kind of sensitive data there, you're screwed. We also need to know the easiest way of storing a token safely is just keeping it in memory, but this option has some pitfalls in UX. So what should we do?

Reviewer: ```suggestion
Before we get started, know that using local storage is not a good idea. Malicious scripts can iterate over local storage and send the info to an external API. So, if you have any kind of sensitive data there, you're screwed.

The easiest way of storing a token safely is by just keeping it in memory. This option has some pitfalls in UX. So what should we do?
```

> Cookies are a good tool for this scenario, but of course you should follow some guidelines to make it safe such as set the cookie as `secure`, `httpOnly` and `SameSite=strict`. The `secure` flag ensures that the cookie is sent only over `https` connections, the `httpOnly` makes the cookie inaccessible on client side so malicious scripts cannot see your sensitive data in `document.cookies` and `SameSite=strict` blocks cookies from being sent to an external domain.

Reviewer: ```suggestion
Cookies are a good tool for this scenario. Of course, you should follow some guidelines to make it safe to do so, such as set the cookie as `secure`, `httpOnly`, and `SameSite=strict`. The `secure` flag ensures that the cookie is sent only over `https` connections, the `httpOnly` flag makes the cookie inaccessible on the client, so malicious scripts cannot see your sensitive data in `document.cookies`, and the `SameSite=strict` flag blocks cookies from being sent to an external domain.
```

> Since the cookie is not available on client side, we need to create some backend logic to fetch the token and store in memory. You can create a BFF(backend for frontends) using a web framework like Rails, Django, Phoenix, Laravel, etc. or you can use lambdas which are very easy and fast to move on. I recommend that you use Netlify Functions or Zeit. Let's see how it looks like.

Reviewer: ```suggestion
Since the cookie is not available on the client, we need to create some backend logic to fetch the token and store it in memory. You can create a BFF(backend for frontends) using a web framework like Rails, Django, Phoenix, Laravel, etc., or you can use lambdas which are very easy and fast to move on. I recommend that you use Netlify Functions or Zeit. Let's see how it works.
```

Reviewer: When you say "store it in memory" here, we're talking about the backend. Why is it important for the backend to store it in memory, and not on disk, in a database, etc? A front-end cannot access the internals of the backend at will: a frontend must use a backend's API to fetch/modify data.

Later in the article, it's evident that the frontend code (the app) is storing the token in memory. If that's the intention, the memory storage in this paragraph needs to be corrected. Also, I would strongly suggest that it is not the app storing the token, but the browser.

If `httpOnly` restricts access from client code, then the frontend code cannot access it at all, correct? When given a cookie by the backend, the browser stores the cookie and includes it in subsequent requests. If this description matches your understanding, then I'd also suggest a different description than "store it in memory". Browsers operationally keep cookies in memory, but they also persist cookies in local file storage.

Author: > Later in the article, it's evident that the frontend code (the app) is storing the token in memory. If that's the intention, the memory storage in this paragraph needs to be corrected. Also, I would strongly suggest that it is not the app storing the token, but the browser.

Both assumptions are right!

> If httpOnly restricts access from client code, then the frontend code cannot access it at all, correct? When given a cookie by the backend, the browser stores the cookie and includes it in subsequent requests. If this description matches your understanding, then I'd also suggest a different description than "store it in memory". Browsers operationally keep cookies in memory, but they also persist cookies in local file storage.

Yeah, makes sense. The intention here when I said "memory" is store it in a run time variable instead of put it on local storage, session storage or cookies.

### Login Flow

![Login Flow](./login-flow.png)

> In this flow what we are doing is creating a token and storing it in a httpOnly cookie to be protected by our server side preventing it to be accessed from malicious front-end/client scripts. After that, the app can stores the token in memory and use it to run your operations. 

Reviewer: ```suggestion
In this flow, we creating a token and storing it in a `httpOnly` cookie, to be protected by our backend, preventing it from being accessed by malicious front-end/client scripts. After that, the app can store the token in memory and use it to run your logic. 
```

Reviewer: When you say "the app can store the token in memory" here, are you referring to the frontend, or the backend? App seems quite ambiguous.

Author: front-end

Author: After saw your comments I'll change that.

> *Notice the function is acting as the cookie guardian where is stored the token.*

Reviewer: ```suggestion
*Notice that the function is acting as the cookie guardian where the token is stored.*
```

> You can see a more detailed explanation bellow:

Reviewer: ```suggestion
Here is a more detailed explanation:
```

> 1. User access the login page.

Reviewer: ```suggestion
1. A user accesses the login page.
```

> 2. Send the credentials to login function.

Reviewer: ```suggestion
2. When the user attempts to authenticate, the credentials are sent to the login function.
```

> 3. The login function validates the credentials and creates a token if the credentials are valid. If it is not, it returns an unauthorized error.

Reviewer: ```suggestion
3. The login function validates the credentials and creates a token if the credentials are valid. If the credentials are not valid, the login function returns an "unauthorized" error.
```

> 4. If it is valid, the function saves the token in the cookie as httpOnly so, only servers can access it preventing it to be accessed from malicious front-end/client scripts.

Reviewer: ```suggestion
4. For valid credentials, the login function creates an `httpOnly` cookie, which prevents access from malicious front-end/client scripts.
```

> 5. (6 and 7) The function returns the token and the app saves it in memory.

Reviewer: The diagram is, in my opinion, somewhat confusing/misleading. You show the cookie as being associated with the backend, as if that's where it lives. A cookie is a header in an HTTP request, which means that the backend communicates the cookie to the frontend, so both contexts have access to the cookie. Yet, step 6 shows that communication not as a cookie but as a token.

I think it would also be helpful to place the memory to the left of the client, because it is the client's memory. Placing it to the right makes it closer to the backend in the diagram, and the backend cannot access the client memory.

Author: Yeah, I'll do that. Thanks.

> Ok, so if my user refreshs the page it will loose the token and crash my app so what we should do? We need to validate the user session trough a validate function.

Reviewer: ```suggestion
One problem: if the user refreshes the page, their browser loses the token and they become unauthenticated, which might crash my app. What we should do? We need to validate the user session through a validate function.
```

Reviewer: Is this correct? Normally, session cookies are invalidated when the browser window is closed, not when the page is refreshed. Permanent cookies have an expiration, but that expiration is not reached merely by refreshing the page.

Author: Yeah, here the concept is when user store the token in a run time variable but it is not right. I'll update that.

### Validation Flow

![Validate Flow](./validate-flow.png)

> In this flow we need to check if the user is already logged in. So the app send a request to the validate function which will check if there is any token stored in the cookie and if it is valid. If it is valid, the validate functions returns it to the app which will store it in the memory. Like we did before, I put it in a more descriptive way right bellow:

Reviewer: ```suggestion
We need to check if the user is already logged in. The app sends a request to the validate function, which checks if there is any token stored in the cookie and if it is valid. If the token is valid, the validate function returns it to the app, which stores the token in memory. Here is a more detailed description:
```

> 1. User access the app.

Reviewer: ```suggestion
1. A user accesses the app.
```

> 2. The app validates the user session.

Reviewer: ```suggestion
2. The app makes a request to the validate function.
```

> 3. (4 and 5) The function check if there is a token and if it is valid. If it is not, it returns an unauthorized error.

Reviewer: ```suggestion
3. (4 and 5) The function checks if an authorization cookie exists, that the cookie contains a token, and whether the token is valid. If none of these things are true, the function returns an "unauthorized" error.
```

> 6. If it is valid, the function returns the token and the app saves it in memory. 

Reviewer: ```suggestion
6. If the token is valid, the function returns the token and the app saves it in memory. 
```

Reviewer: Why does the function return the token? The client provided the token, and it is valid, so the client already has a valid token. Does the token change during validation? Is this performing the function of a nonce as well?

Author: You are right. The function does not need to return the token. I'll change that, thanks for the call.

> Now we know how to improve your security storing token safely we can implement it. In the next article we are going to implement it using JS, React and FaunaDB. I hope this article is helpful to you and I'll let some nice references bellow for further reading.

Reviewer: ```suggestion
Now that we know how to improve your security by storing a token safely, we can implement it. In the next article, I'll show you how using JS, React and FaunaDB. I hope this article has been helpful to you. For further information, check out the following references:
```
